### ISSUE TITLE
FTP timeout configuration

### ISSUE THREAD

**Issue by noff** [backup/backup#700]
Hi.
It will be great to have FTP timeout configuration because sometimes connection delays happened and backup interrupts before successfully uploaded.

**Comment by tombruijn** [comment]
Hi @noff, congrats on opening issue number 700!

I'm not that familiar with the [Net::FTP](http://ruby-doc.org/stdlib-2.2.3/libdoc/net/ftp/rdoc/Net/FTP.html). Looks like it's part of ruby itself, unlike [Net::SFTP]().

I see that the documentation reference a [open_timeout](http://ruby-doc.org/stdlib-2.2.3/libdoc/net/ftp/rdoc/Net/FTP.html#open_timeout) and a [read_timeout](http://ruby-doc.org/stdlib-2.2.3/libdoc/net/ftp/rdoc/Net/FTP.html#read_timeout). Which setting do you need? Or do you think we'd have to configure both? Separately even?

I see for Net::SFTP (and Net::SSH) we'd only have to configure [timeout](http://net-ssh.github.io/net-ssh/classes/Net/SSH.html#method-c-start).

**Comment by noff** [comment]
Hi. It's to prevent this error:

```
[2015/09/22 05:22:40][error] Model::Error: Backup for modelname
Failed!
[2015/09/22 05:22:40][error] --- Wrapped Exception ---
[2015/09/22 05:22:40][error] Net::FTPTempError: 426 Transfer aborted.  Data connection closed.
[2015/09/22 05:22:40][error] 
[2015/09/22 05:22:40][error] Backtrace:
[2015/09/22 05:22:40][error]   /home/rails/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/net/ftp.rb:325:in `getresp'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/net/ftp.rb:563:in `rescue in storbinary'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/net/ftp.rb:542:in `storbinary'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/net/ftp.rb:694:in `putbinaryfile'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/net/ftp.rb:722:in `put'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/storage/ftp.rb:58:in `block (2 levels) in transfer!'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/storage/ftp.rb:54:in `each'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/storage/ftp.rb:54:in `block in transfer!'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/storage/ftp.rb:46:in `block in connection'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/net/ftp.rb:135:in `open'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/storage/ftp.rb:44:in `connection'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/storage/ftp.rb:51:in `transfer!'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/storage/base.rb:35:in `perform!'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/model.rb:334:in `block in store!'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/model.rb:332:in `map'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/model.rb:332:in `store!'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/model.rb:306:in `block in procedures'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/model.rb:269:in `call'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/model.rb:269:in `block in perform!'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/model.rb:268:in `each'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/model.rb:268:in `perform!'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/lib/backup/cli.rb:156:in `perform'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/thor-0.18.1/lib/thor/command.rb:27:in `run'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/thor-0.18.1/lib/thor/invocation.rb:120:in `invoke_command'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/thor-0.18.1/lib/thor.rb:363:in `dispatch'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/thor-0.18.1/lib/thor/base.rb:439:in `start'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/gems/backup-4.1.11/bin/backup:5:in `<top (required)>'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/bin/backup:23:in `load'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/bin/backup:23:in `<main>'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/bin/ruby_executable_hooks:15:in `eval'
[2015/09/22 05:22:40][error]   /home/rails/.rvm/gems/ruby-2.2.0/bin/ruby_executable_hooks:15:in `<main>'
```

Storage support confirms problem on their side and suggests to increase timeout as temporary solution. I think it's open_timeout but it is better to configure both.

**Comment by tombruijn** [comment]
@noff I quickly added the option to configure the timeout, see this PR: https://github.com/backup/backup/pull/702

Can you try installing backup from this branch and get back to me if it helps?

**Comment by noff** [comment]
Hi. Sorry for long time to answer.
How can I install it using "gem install ..." command?

**Comment by tombruijn** [comment]
Using bundler you can add this to your Gemfile

```ruby
gem "backup", :github => "backup/backup"
#or
gem "backup", :git => "https://github.com/backup/backup.git"
```

and run bundle install

**Comment by noff** [comment]
Did this:

```
cd ~/Backup
vi Gemfile
# added gem string
bundle
backup perform -t dbtables

===>

$ backup perform -t dbtables
[2015/10/02 09:36:22][error] CLI::Error
[2015/10/02 09:36:22][error] --- Wrapped Exception ---
[2015/10/02 09:36:22][error] NoMethodError: undefined method open_timeout= for #<Backup::Storage::FTP:0x000000039c6e38>
[2015/10/02 09:36:22][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config/helpers.rb:122:in `method_missing'
```

**Comment by tombruijn** [comment]
Sorry @noff , should have provided an example. I named the option just `timeout`

```ruby
store_with FTP do |ftp|
  ftp.timeout = 10
end
```

**Comment by noff** [comment]
```
$ /bin/bash -l -c 'cd /home/rails/Backup && backup perform -t dbtables --quiet'
[2015/10/02 20:53:19][error] CLI::Error
[2015/10/02 20:53:19][error] --- Wrapped Exception ---
[2015/10/02 20:53:19][error] NoMethodError: undefined method `timeout=' for #<Backup::Storage::FTP:0x00000003ad2e58>
[2015/10/02 20:53:19][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config/helpers.rb:122:in `method_missing'
[2015/10/02 20:53:19][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config/helpers.rb:86:in `block in load_defaults!'
[2015/10/02 20:53:19][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config/helpers.rb:83:in `each'
[2015/10/02 20:53:19][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config/helpers.rb:83:in `load_defaults!'
[2015/10/02 20:53:19][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/storage/base.rb:39:in `initialize'
[2015/10/02 20:53:19][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/storage/ftp.rb:22:in `initialize'
[2015/10/02 20:53:19][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/model.rb:153:in `new'
[2015/10/02 20:53:19][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/model.rb:153:in `store_with'
```

**Comment by tombruijn** [comment]
@noff you'll need to execute it in the bundler context. Run it with `bundle exec backup ...` otherwise it will use the normal installation of the backup gem. Please try it with bundler and let me know if it works

**Comment by noff** [comment]
```
$ bundle exec backup perform -t dbtables
[2015/10/04 16:46:05][error] CLI::Error
[2015/10/04 16:46:05][error] --- Wrapped Exception ---
[2015/10/04 16:46:05][error] NoMethodError: undefined method `timeout=' for #<Backup::Storage::FTP:0x00000004872c98>
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config/helpers.rb:122:in `method_missing'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config/helpers.rb:86:in `block in load_defaults!'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config/helpers.rb:83:in `each'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config/helpers.rb:83:in `load_defaults!'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/storage/base.rb:39:in `initialize'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/storage/ftp.rb:22:in `initialize'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/model.rb:153:in `new'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/model.rb:153:in `store_with'
[2015/10/04 16:46:05][error] /home/rails/Backup/models/shard00.rb:22:in `block (2 levels) in load'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/model.rb:127:in `instance_eval'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/model.rb:127:in `initialize'
[2015/10/04 16:46:05][error] /home/rails/Backup/models/shard00.rb:4:in `new'
[2015/10/04 16:46:05][error] /home/rails/Backup/models/shard00.rb:4:in `block in load'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config.rb:48:in `instance_eval'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config.rb:48:in `block in load'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config.rb:47:in `each'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/config.rb:47:in `load'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/lib/backup/cli.rb:129:in `perform'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/gems/thor-0.18.1/lib/thor/command.rb:27:in `run'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/gems/thor-0.18.1/lib/thor/invocation.rb:120:in `invoke_command'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/gems/thor-0.18.1/lib/thor.rb:363:in `dispatch'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/gems/thor-0.18.1/lib/thor/base.rb:439:in `start'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bundler/gems/backup-4057a4b2dac8/bin/backup:5:in `<top (required)>'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bin/backup:23:in `load'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bin/backup:23:in `<main>'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bin/ruby_executable_hooks:15:in `eval'
[2015/10/04 16:46:05][error] /home/rails/.rvm/gems/ruby-2.2.3/bin/ruby_executable_hooks:15:in `<main>'
```

I use default settings

```
   Storage::FTP.defaults do |server|
    ...
    server.keep       = 5
    server.passive_mode = true
    server.timeout = 240
  end
```

**Comment by tombruijn** [comment]
@noff sorry, I'm an idiot!

I forgot to add the ref to the code you have to add to your gemfile. Please use the code below in your gemfile and try again

```ruby
gem "backup", :git => "https://github.com/backup/backup.git", ref: "ftp_timeout_config"
```

**Comment by noff** [comment]
It works now. Thank you.

**Comment by tombruijn** [comment]
@noff awesome! Let me know in a couple days if you have or have not encountered any problems. I'll also try to add the functionality to the sftp setup (if necessary and possible) before merging it and publishing a new version

**Comment by tombruijn** [comment]
This was released today in backup 4.2.1

**Comment by noff** [comment]
Great, thanks.
